src/libmongoc/tests/json-test.c

```diff
@@ -176,47 +176,149 @@ server_description_by_hostname (mongoc_topology_description_t *topology,
+         bson_iter_init_find (&ismaster_field_iter, &ismaster, "1");
+         bson_iter_bson (&ismaster_field_iter, &response);
+
+         /* send ismaster through the topology description's handler */
+         capture_logs (true);
+         mongoc_topology_description_handle_ismaster (
+            td, sd->id, &response, 1, NULL);
+         if (td->servers->items_len == 0) {
+            ASSERT_CAPTURED_LOG ("topology",
+                                 MONGOC_LOG_LEVEL_WARNING,
+                                 "Last server removed from topology");
+         }
+         capture_logs (false);
+      }
+   } else if (bson_iter_init_find (
+                 &phase_field_iter, phase, "applicationErrors")) {
+      bson_t app_errors;
+      bson_iter_t app_error_iter;
+      bson_t app_error;
+      bson_iter_t app_error_field_iter;
+
+      bson_iter_bson (&phase_field_iter, &app_errors);
+      bson_iter_init (&app_error_iter, &app_errors);
+
+      while (bson_iter_next (&app_error_iter)) {
+         uint32_t generation = 0;
+         uint32_t max_wire_version = 0;
+         const char *when_str;
+         _mongoc_sdam_app_error_when_t when = 0;
+         const char *type_str;
+         _mongoc_sdam_app_error_type_t type = 0;
+         bson_t response;
+         bson_error_t err;
+
+         bson_iter_bson (&app_error_iter, &app_error);
+
+         bson_iter_init_find (&app_error_field_iter, &app_error, "address");
+         hostname = bson_iter_utf8 (&app_error_field_iter, NULL);
+         sd = server_description_by_hostname (td, hostname);
+
+         /* if server has been removed from topology, skip */
+         if (!sd) {
+            continue;
+         }
 
-      bson_iter_init_find (&ismaster_field_iter, &ismaster, "0");
-      hostname = bson_iter_utf8 (&ismaster_field_iter, NULL);
-      sd = server_description_by_hostname (td, hostname);
+         if (bson_iter_init_find (
+                &app_error_field_iter, &app_error, "generation")) {
+            BSON_ASSERT (BSON_ITER_HOLDS_INT32 (&app_error_field_iter));
+            generation = bson_iter_int32 (&app_error_field_iter);
+         } else {
+            /* Default to the current generation. */
+            generation = sd->generation;
+         }
 
-      /* if server has been removed from topology, skip */
-      if (!sd) {
-         continue;
-      }
+         BSON_ASSERT (bson_iter_init_find (
+            &app_error_field_iter, &app_error, "maxWireVersion"));
+         BSON_ASSERT (BSON_ITER_HOLDS_INT32 (&app_error_field_iter));
+         max_wire_version = bson_iter_int32 (&app_error_field_iter);
+
+         BSON_ASSERT (
+            bson_iter_init_find (&app_error_field_iter, &app_error, "when"));
+         BSON_ASSERT (BSON_ITER_HOLDS_UTF8 (&app_error_field_iter));
+         when_str = bson_iter_utf8 (&app_error_field_iter, 0);
+         if (0 == strcmp (when_str, "beforeHandshakeCompletes")) {
+            when = MONGOC_SDAM_APP_ERROR_BEFORE_HANDSHAKE;
```

Comment:
Is this the only appearance of `MONGOC_SDAM_APP_ERROR_BEFORE_HANDSHAKE` in the entire PR? I only recall seeing `MONGOC_SDAM_APP_ERROR_AFTER_HANDSHAKE` elsewhere.

Is that an oversight if so? I would have assumed there'd be at least once instance of the "before" case in the driver.

Reply:
Ah, I wasn't thinking that the enum was being used for documentation/clarity purposes. That's great so no concern here.